libs/application-generic/src/services/feature-flags/types.ts

```diff
@@ -3,12 +3,16 @@ import { FeatureFlagsKeysEnum } from '@novu/shared';
 
 type PartialWithId<T> = Partial<T> & { _id: string };
 
-export type FeatureFlagContextBase = {
-  environment?: PartialWithId<EnvironmentEntity>;
-  organization?: PartialWithId<OrganizationEntity>;
-  user?: PartialWithId<UserEntity>;
+type ContextEntities = {
+  environment: PartialWithId<EnvironmentEntity>;
+  organization: PartialWithId<OrganizationEntity>;
+  user: PartialWithId<UserEntity>;
 };
 
+type RequireAtLeastOne<T> = { [K in keyof T]: { [P in K]: T[P] } & { [P in Exclude<keyof T, K>]?: T[P] } }[keyof T];
+
+export type FeatureFlagContextBase = RequireAtLeastOne<ContextEntities>;
```

Comment:
another solution could be 

```suggestion
export type FeatureFlagContextBase =
  | {
      environment: PartialWithId<EnvironmentEntity>;
      organization?: PartialWithId<OrganizationEntity>;
      user?: PartialWithId<UserEntity>;
    }
  | {
      environment?: PartialWithId<EnvironmentEntity>;
      organization: PartialWithId<OrganizationEntity>;
      user?: PartialWithId<UserEntity>;
    }
  | {
      environment?: PartialWithId<EnvironmentEntity>;
      organization?: PartialWithId<OrganizationEntity>;
      user: PartialWithId<UserEntity>;
    };
type ContextEntities = {
  environment: PartialWithId<EnvironmentEntity>;
  organization: PartialWithId<OrganizationEntity>;
  user: PartialWithId<UserEntity>;
};```

Reply:
I prefer the discriminated unions but I will leave it to you :)